### PR TITLE
add more slots to the processing pattern grid

### PR DESCRIPTION
**Pull request by Darkere**
This implements more slots for the processing grid. 
The number of slots is defined in `GridNetworkNode#processingmatrixsize` and that value or the size of the Item/Fluid-handler is used everywhere. 
It should be possible to make that number a config even. I've set it to 81 for a 9x9 crafting table. 

I've set it up so that the scroll bar unlocks as more slots are filled, making sure that always at least 1 item is visible to prevent confusion caused by accidental scrolling. This also creates an indicator in case an item was accidentally left further down in the recipe. 

Could make screenshots but I figure you want to take a look anyway :)

### CODE REVIEW

**Comment by raoulvdberge** [comment]
Can this be updated for 1.16?

**Comment by Darkere** [comment]
Ah yeah. Will do

**Comment by Darkere** [comment]
Should be good to go. 

There is one uglyness with that isItems boolean that I can't seem to fix. 
But I think the plan is to get rid of item fluid switching which would make it obsolete.

File: src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/GridNetworkNode.java [hunk 1]
```diff
@@ -77,7 +77,9 @@
     private static final String NBT_PROCESSING_MATRIX_FLUIDS = "ProcessingMatrixFluids";
     private static final String NBT_ALLOWED_TAGS = "AllowedTags";
 
-    private final AllowedTagList allowedTagList = new AllowedTagList(this::updateAllowedTags);
+    public static int processingMatrixSize = 81;
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/apiimpl/network/node/GridNetworkNode.java:80]: final, PROCESSING_MATRIX_SIZE

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 2]
```diff
@@ -128,12 +140,24 @@ public void onPostInit(int x, int y) {
         addSideButton(new GridSizeSideButton(this, () -> grid.getSize(), size -> grid.onSizeChanged(size)));
 
         if (grid.getGridType() == GridType.PATTERN) {
+
+            patternScrollbar = new ScrollbarWidget(this, 160, getTopHeight() + getVisibleRows() * 18 + 4, 6, 18 * 3 - 2, true);
+            patternScrollOffsetAbsoluteMax = GridNetworkNode.processingMatrixSize / 3 - 3;
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:145]: Why is this an instance variable?

File: src/main/java/com/refinedmods/refinedstorage/container/GridContainer.java [hunk 3]
```diff
@@ -288,4 +306,24 @@ public boolean canMergeSlot(ItemStack stack, Slot slot) {
...
-}
+
+    public void updatePatternSlotPositions(int newOffset) {
+        patternScrollOffset = newOffset;
+        int yPosition = screenInfoProvider.getTopHeight() + (screenInfoProvider.getVisibleRows() * 18) + 4;
+
+        for (int i = 0; i < itemPatternSlots.size(); i++) {
+
+            if (i == GridNetworkNode.processingMatrixSize) { // reset when reaching output slots
+                yPosition = screenInfoProvider.getTopHeight() + (screenInfoProvider.getVisibleRows() * 18) + 4;
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/container/GridContainer.java:317]: yPosition = originalYPosition

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 4]
```diff
@@ -562,6 +628,47 @@ public void updateExactPattern(boolean checked) {
...
+                    filledOutputSlots++;
+                    lastFilledOutputSlot = i - GridNetworkNode.processingMatrixSize - 1;
+                } else {
+                    filledInputSlots++;
+                    lastFilledInputSlot = i;
+                }
+
+            }
+        }
+        patternScrollOffsetMax = Math.max(Math.floorDiv(Math.max(filledInputSlots, filledOutputSlots) - 1, 3),
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:650]: This line is unreadable

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 5]
```diff
@@ -513,19 +560,38 @@ private boolean isOverCraftingOutputArea(double mouseX, double mouseY) {
...
 
+        if (grid.getGridType() == GridType.PATTERN) {
+            patternScrollbar.mouseMoved(mx, my);
+        }
         super.mouseMoved(mx, my);
     }
 
     @Override
     public boolean mouseReleased(double mx, double my, int button) {
+        if (grid.getGridType() == GridType.PATTERN) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:571]: two ifs = 1 if

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 6]
```diff
@@ -513,19 +560,38 @@ private boolean isOverCraftingOutputArea(double mouseX, double mouseY) {
...
+            if (patternScrollbar.mouseReleased(mx, my, button)) {
+                return true;
+            }
+        }
         return scrollbar.mouseReleased(mx, my, button) || super.mouseReleased(mx, my, button);
     }
 
     @Override
     public boolean mouseScrolled(double x, double y, double delta) {
+        if (grid.getGridType() == GridType.PATTERN) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:581]: 3 ifs = 1 if

**Review by raoulvdberge — CHANGES_REQUESTED** [review]

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 7]
```diff
@@ -326,6 +364,11 @@ public void renderBackground(MatrixStack matrixStack, int x, int y, int mouseX,
             }
 
             blit(matrixStack, x + 172, y + getTopHeight() + (getVisibleRows() * 18) + 22, 240, ty * 16, 16, 16);
+            if (processingPattern.isChecked()) {
+                updatePatternScrollbar();
+                patternScrollbar.render(matrixStack);
+
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:370]: Useless newline....

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 8]
```diff
@@ -128,12 +140,24 @@ public void onPostInit(int x, int y) {
         addSideButton(new GridSizeSideButton(this, () -> grid.getSize(), size -> grid.onSizeChanged(size)));
 
         if (grid.getGridType() == GridType.PATTERN) {
+
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:143]: Useless newline

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 9]
```diff
@@ -128,12 +140,24 @@ public void onPostInit(int x, int y) {
...
+            patternScrollbar = new ScrollbarWidget(this, 160, getTopHeight() + getVisibleRows() * 18 + 4, 6, 18 * 3 - 2, true);
+            patternScrollOffsetAbsoluteMax = GridNetworkNode.processingMatrixSize / 3 - 3;
+            patternScrollbar.setMaxOffset(patternScrollOffsetAbsoluteMax);
+            patternScrollbar.setOffset(patternScrollOffset); // keep offset when changing between fluid and item view
+            container.updatePatternSlotPositions(patternScrollOffset);
+            patternScrollbar.addListener((oldOffset, newOffset) -> {
+                patternScrollOffset = newOffset;
+                container.updatePatternSlotPositions(newOffset);
+            });
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:152]: This block of code needs some more newlines here and there. It's quite unreadable now.

File: src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java [hunk 10]
```diff
@@ -428,7 +471,11 @@ public boolean mouseClicked(double mouseX, double mouseY, int clickedButton) {
         if (scrollbar.mouseClicked(mouseX, mouseY, clickedButton)) {
             return true;
         }
-
+        if (grid.getGridType() == GridType.PATTERN) {
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/grid/GridScreen.java:474]: 2 ifs = 1 if

File: src/main/java/com/refinedmods/refinedstorage/screen/widget/ScrollbarWidget.java [hunk 11]
```diff
@@ -62,7 +67,7 @@ public void render(MatrixStack matrixStack) {
         RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
 
         screen.bindTexture(RS.ID, "icons.png");
-        screen.blit(matrixStack, screen.getGuiLeft() + x, screen.getGuiTop() + y + (int) Math.min(height - SCROLLER_HEIGHT, (float) offset / (float) maxOffset * (float) (height - SCROLLER_HEIGHT)), isEnabled() ? 232 : 244, 0, 12, 15);
+        screen.blit(matrixStack, screen.getGuiLeft() + x, screen.getGuiTop() + y + (int) Math.min(height - SCROLLER_HEIGHT, (float) offset / (float) maxOffset * (float) (height - SCROLLER_HEIGHT)),small ? (isEnabled() ? 218 : 225) : (isEnabled() ? 232 : 244), 0, small ? 7 : 12, SCROLLER_HEIGHT);
```
Comment (raoulvdberge) [inline src/main/java/com/refinedmods/refinedstorage/screen/widget/ScrollbarWidget.java:70]: These changes make the line unreadable. Please move these to variables.

**Comment by raoulvdberge** [comment]
Did a first pass. Lots of formatting issues, please reformat.

**Pull request merged**